File: src/dynamic_programming/cpp/can_construct/test/test_can_construct.cpp
#include "can_construct.h"
#include "gtest/gtest.h"
#include <algorithm>

TEST(CanConstructTest, TestNegativeBasic1) {
  std::vector<std::string> wordBank = {"bo",  "rd", "ate", "t",
                                       "ska", "sk", "boar"};
  std::string target = "skateboard";

  ASSERT_FALSE(canConstructBasic(target, wordBank));
}

TEST(CanConstructTest, TestNegativeBasic2) {
  std::vector<std::string> wordBank = {"mo", "ha", "cz"};
  std::string target = "mohach";

  ASSERT_FALSE(canConstructBasic(target, wordBank));
}

TEST(CanConstructTest, TestPositiveBasic1) {
  std::vector<std::string> wordBank = {"a", "b", "c"};
  std::string target = "abc";

  ASSERT_TRUE(canConstructBasic(target, wordBank));
}

TEST(CanConstructTest, TestPositiveBasic2) {
  std::vector<std::string> wordBank = {"ab", "abc", "cd", "def", "abcd"};
  std::string target = "abcdef";

  ASSERT_TRUE(canConstructBasic(target, wordBank));
}

TEST(CanConstructTest, TestNegativeMemo1) {
  std::vector<std::string> wordBank = {"bo",  "rd", "ate", "t",
                                       "ska", "sk", "boar"};
  std::string target = "skateboard";

  ASSERT_FALSE(canConstructBasic(target, wordBank));
}

TEST(CanConstructTest, TestNegativeMemo2) {
  std::vector<std::string> wordBank = {"mo", "ha", "cz"};
  std::string target = "mohach";

  ASSERT_FALSE(canConstructBasic(target, wordBank));
}

TEST(CanConstructTest, TestPositiveMemo1) {
  std::vector<std::string> wordBank = {"a", "b", "c"};
  std::string target = "abc";

  ASSERT_TRUE(canConstructBasic(target, wordBank));
}

TEST(CanConstructTest, TestPositiveMemo2) {
  std::vector<std::string> wordBank = {"ab", "abc", "cd", "def", "abcd"};
  std::string target = "abcdef";

  ASSERT_TRUE(canConstructBasic(target, wordBank));
}

TEST(CanConstructTest, TestNegativeTable1) {
  std::vector<std::string> wordBank = {"bo",  "rd", "ate", "t",
                                       "ska", "sk", "boar"};
  std::string target = "skateboard";

  ASSERT_FALSE(canConstructBasic(target, wordBank));
}

TEST(CanConstructTest, TestNegativeTable2) {
  std::vector<std::string> wordBank = {"mo", "ha", "cz"};
  std::string target = "mohach";

  ASSERT_FALSE(canConstructBasic(target, wordBank));
}

TEST(CanConstructTest, TestPositiveTable1) {
  std::vector<std::string> wordBank = {"a", "b", "c"};
  std::string target = "abc";

  ASSERT_TRUE(canConstructBasic(target, wordBank));
}

TEST(CanConstructTest, TestPositiveTable2) {
  std::vector<std::string> wordBank = {"ab", "abc", "cd", "def", "abcd"};
  std::string target = "abcdef";

  ASSERT_TRUE(canConstructBasic(target, wordBank));
}

File: src/dynamic_programming/python/can_construct/tests/test_can_construct.py
import unittest


from src.can_construct import (
    can_construct_basic,
    can_construct_memo,
    can_construct_table,
)


class TestCanConstructBasic(unittest.TestCase):
    def test_negative(self):
        word_bank = ["bo", "rd", "ate", "t", "ska", "sk", "boar"]
        target = "skateboard"
        self.assertFalse(can_construct_basic(target, word_bank))

        word_bank = ["mo", "ha", "cz"]
        target = "mocha"
        self.assertFalse(can_construct_basic(target, word_bank))

    def test_positive(self):
        word_bank = ["a", "b", "c"]
        target = "abc"
        self.assertTrue(can_construct_basic(target, word_bank))

        word_bank = ["ab", "abc", "cd", "def", "abcd"]
        target = "abcdef"
        self.assertTrue(can_construct_basic(target, word_bank))


class TestCanConstructMemo(unittest.TestCase):
    def test_negative(self):
        word_bank = ["bo", "rd", "ate", "t", "ska", "sk", "boar"]
        target = "skateboard"
        self.assertFalse(can_construct_memo(target, word_bank))

        word_bank = ["mo", "ha", "cz"]
        target = "mocha"
        self.assertFalse(can_construct_memo(target, word_bank))

    def test_positive(self):
        word_bank = ["a", "b", "c"]
        target = "abc"
        self.assertTrue(can_construct_memo(target, word_bank))

        word_bank = ["ab", "abc", "cd", "def", "abcd"]
        target = "abcdef"
        self.assertTrue(can_construct_memo(target, word_bank))


class TestCanConstructTab(unittest.TestCase):
    def test_negative(self):
        word_bank = ["bo", "rd", "ate", "t", "ska", "sk", "boar"]
        target = "skateboard"
        self.assertFalse(can_construct_table(target, word_bank))

        word_bank = ["mo", "ha", "cz"]
        target = "mocha"
        self.assertFalse(can_construct_table(target, word_bank))

    def test_positive(self):
        word_bank = ["a", "b", "c"]
        target = "abc"
        self.assertTrue(can_construct_table(target, word_bank))

        word_bank = ["ab", "abc", "cd", "def", "abcd"]
        target = "abcdef"
        self.assertTrue(can_construct_table(target, word_bank))


if __name__ == "__main__":
    unittest.main()
